### PR TITLE
feat(Chain): move check_refcount_map into Store Validator + caching Tx Refcounts while GCing

### PR DESCRIPTION
**Pull request by Kouprin**
- move check_refcount_map into Store Validator
- caching Tx Refcounts while GCing
- fixes `sanity/epoch_switches.py`
- validation of State Headers and State Parts
- refactoring

TEST PLAN
---------
CI + custom Nightly

### CODE REVIEW

File: chain/client/tests/process_blocks.rs [hunk 1]
```diff
@@ -875,7 +873,6 @@ fn test_gc_block_skips() {
             env.produce_block(0, i);
         }
     }
-    assert!(check_refcount_map(&mut env.clients[0].chain).is_ok());
```
Comment (bowenwang1996) [inline chain/client/tests/process_blocks.rs:878]: why are these checks removed?

File: chain/jsonrpc/src/lib.rs [hunk 2]
```diff
@@ -753,20 +752,6 @@ impl JsonRpcHandler {
         }
     }
 
-    async fn adv_check_refmap(&self, _params: Option<Value>) -> Result<Value, RpcError> {
```
Comment (bowenwang1996) [inline chain/jsonrpc/src/lib.rs:756]: why is this removed?

File: chain/chain/src/store.rs [hunk 3]
```diff
@@ -2154,17 +2159,12 @@ impl<'a> ChainStoreUpdate<'a> {
     }
 
     pub fn gc_col_transaction(&mut self, tx_hash: CryptoHash) -> Result<(), Error> {
-        let mut refs = self.get_tx_refcount(&tx_hash)?;
-        debug_assert!(refs > 0);
-        refs = refs.saturating_sub(1);
-        if refs == 0 {
+        let mut refcount = self.get_tx_refcount(&tx_hash)?;
+        refcount -= 1;
```
[the remark this answers is not in the document]
Comment (Kouprin) [reply chain/chain/src/store.rs:2163]: Storage must be consistent.

File: chain/client/tests/process_blocks.rs (hunk 1 above)
Comment (Kouprin) [reply chain/client/tests/process_blocks.rs:878]: They were added to sanity tests when there were not enough coverage of python tests and no Store Validator introduced. We don't need those checks anymore.

File: chain/jsonrpc/src/lib.rs (hunk 2 above)
Comment (Kouprin) [reply chain/jsonrpc/src/lib.rs:756]: Because Block Refmap validation is moved into Store Validator as one of many ways to check Storage Consistency. Running Store Validator you run checking Block Refmap as well.

File: chain/chain/src/store.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Kouprin) [reply chain/chain/src/store.rs:2169]: Added Err for negative refcount to avoid panic.

File: chain/client/tests/process_blocks.rs (hunk 1 above)
Comment (frol) [reply chain/client/tests/process_blocks.rs:878]: I would still leave such asserts in the code, at least as `debug_assert!`. This can serve the documentation purpose and potentially help with refactorings if any will happen in the future.

File: chain/chain/src/store.rs [hunk 4]
```diff
@@ -2393,32 +2399,25 @@ impl<'a> ChainStoreUpdate<'a> {
...
             };
+
             // Increase transaction refcounts for all included txs
             for tx in chunk.transactions.iter() {
-                let mut refs = self.chain_store.get_tx_refcount(&tx.get_hash())?;
-                refs += 1;
-                store_update.set_ser(ColTransactionRefCount, tx.get_hash().as_ref(), &refs)?;
+                let mut refcount = self.chain_store.get_tx_refcount(&tx.get_hash())?;
+                refcount += 1;
+                store_update.set_ser(ColTransactionRefCount, tx.get_hash().as_ref(), &refcount)?;
```
Comment (frol) [inline chain/chain/src/store.rs:2407]: Do we ensure that this does not lead us to a race condition?

File: chain/client/tests/process_blocks.rs (hunk 1 above)
Comment (Kouprin) [reply chain/client/tests/process_blocks.rs:878]: there is no 'check_refcount_map' anymore

File: chain/chain/src/store.rs (hunk 4 above)
Comment (Kouprin) [reply chain/chain/src/store.rs:2407]: We cannot increase and decrease Tx Refcount in one store update operation. While increasing, we cannot add same Tx twice. There is also a comment in `finalize()` method explaining this.
